### PR TITLE
feat(aggregation): Add MoCo

### PR DESCRIPTION
**Pull request by rkhosrowshahi**
- Linked to #665

Adds MoCo ([ICLR 2023](https://openreview.net/forum?id=dLAYGdKTi2)) as `MoCo` / `MoCoWeighting` with `MatrixWeighting` and `Stateful`, plus docs and tests. Keeps a moving Jacobian estimate $$Y$$ and task weights $$\lambda$$; returns $$\lambda^\top Y$$. Call `reset` between runs.

**Tests:**
- `tests/unit/aggregation/test_moco.py` (behavior, reset, getters/setters, edge cases, alignment with `MoCoWeighting` and `y`)
- `tests/unit/aggregation/test_values.py` (fixed baselines for `MoCo` / `MoCoWeighting`)

### CODE REVIEW

**Comment by rkhosrowshahi** [comment]
Hi @ValerianRey,
Could you please review the code and assign labels?
Thanks!

File: src/torchjd/aggregation/_moco.py [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+        self._y: Tensor | None = None
+        self._lambd: Tensor | None = None
+        self._state_key: tuple[int, int, torch.device, torch.dtype] | None = None
+
+    def forward(self, matrix: Matrix, /) -> Tensor:
+        if matrix.shape[0] == 0:
+            self.reset()
+            self._y = matrix.detach().clone()
+            self._state_key = (matrix.shape[0], matrix.shape[1], matrix.device, matrix.dtype)
+            return matrix.new_empty((0,))
```
Comment (PierreQuinton) [inline src/torchjd/aggregation/_moco.py:118]: This should never happen (I think we can safely assume that `matrix.shape[0] > 0` in the forward call to an aggregator. Maybe we want to specify that somewhere @ValerianRey

File: src/torchjd/aggregation/_moco.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+        self._ensure_state(matrix)
+        self.step += 1
+
+        y = cast(Tensor, self._y)
+        lambd = cast(Tensor, self._lambd)
+
+        beta_step = self.beta / (self.step**self.beta_sigma)
+        gamma_step = self.gamma / (self.step**self.gamma_sigma)
+
+        with torch.no_grad():
```
Comment (PierreQuinton) [inline src/torchjd/aggregation/_moco.py:129]: This is a bit weird, why not having a no_grad around the whole function instead? Also what is the current philosphy on grads of outputs of aggregators @ValerianRey ? Should we unify this?

File: src/torchjd/aggregation/_moco.py (hunk 1 above)
Comment (ValerianRey) [reply src/torchjd/aggregation/_moco.py:118]: I agree, let's get rid of this. And obviously matrices should have at least 1 row, it's not even like autojac can output jacobians with 0 rows. So it would be when a user would really want to aggregate their own empty matrix with this specific aggregator. This will not happen.

File: src/torchjd/aggregation/_moco.py (hunk 2 above)
Comment (ValerianRey) [reply src/torchjd/aggregation/_moco.py:129]: There are just two cases:
1) The aggregator/weighting is differentiable: and we do not want to use any torch.no_grad(), so that the graph is correctly built and stored

2) The aggregator/weighting is non-differentiable (could be for example because some operations are made on numpy arrays, like for UPGrad, DualProj, and a few others, or for some other niche reasons, like PCGrad, and I think similarly GradVac): we want to raise an error when we try to backward through it (we do that already) and I even think we would like to prevent graph construction by wrapping the forward in a torch.no_grad() (which we don't do currently). And now that I think of it, I don't even think we need to raise an error when calling backward on a non-differentiable module if its forward is wrapped in a torch.no_grad(). No graph will ever be created to begin with, so autograd will never try to backward through the module.

So I think that we should improve on that:
- Differentiable aggregators/weighting should inherit from a Differentiable mixin. It will not do anything, and it wont be public, but it will serve as internal documentation.
- Non-differentiable aggregators/weighting should inherit from a NonDifferentiable mixin (similarly protected). It could maybe wrap the forward pass in a torch.no_grad, (and maybe still make it raise an error if we try to differentiate through it).

About MoCo, idk if it's differentiable or not. I guess that changing the value of y inplace will lead to runtime error when we try to differentiate through it (the same reason why we made PCGrad non-differentiable). Need to verify that. If it's not, I think we should make it inherit NonDifferentiable. If it is, let's remove the torch.no_grad and matrix.detach. In any case, let's remove those two things because NonDifferentiable will handle it itself if we add it.

I'll ask claude to work on that.

**Comment by ValerianRey** [comment]
/opencode:Plan please review this and compare this implementation to that of LibMTL (https://github.com/median-research-group/LibMTL/blob/main/LibMTL/weighting/MoCo.py)

**Comment by ValerianRey** [comment]
> Add gradient normalization to match LibMTL/paper

Let's not do that. We will add normalization as wrappers.

File: src/torchjd/aggregation/_moco.py (hunk 2 above)
Comment (ValerianRey) [reply src/torchjd/aggregation/_moco.py:129]: Fixed in #677

Comment (PierreQuinton) [reply src/torchjd/aggregation/_moco.py:129]: Fantastic, upon merge of #677 this should be removed, and then the aggregator/weighting pair should implement the Non differentiable mixin. @rkhosrowshahi Take a look at #677 as it specifically needs to be inherited first.

File: src/torchjd/aggregation/_moco.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+            y = y - beta_step * (y - matrix.detach())
+            yy_t = y @ y.T
+            if self.rho != 0.0:
+                eye = torch.eye(matrix.shape[0], device=matrix.device, dtype=matrix.dtype)
+                yy_t = yy_t + self.rho * eye
+            lambd = torch.softmax(lambd - gamma_step * (yy_t @ lambd), dim=-1)
```
Comment (PierreQuinton) [inline src/torchjd/aggregation/_moco.py:135]: In the paper, the softmax is supposed to be a projection to the probability simplex. Why is this a softmax (it is on the probability simplex, but not a projection)?

Maybe we should email the authors to ask for a clarification on the intended method to project.

File: src/torchjd/aggregation/_moco.py [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+        self.step += 1
+
+        y = cast(Tensor, self._y)
+        lambd = cast(Tensor, self._lambd)
+
+        beta_step = self.beta / (self.step**self.beta_sigma)
+        gamma_step = self.gamma / (self.step**self.gamma_sigma)
+
+        with torch.no_grad():
+            y = y - beta_step * (y - matrix.detach())
```
Comment (PierreQuinton) [inline src/torchjd/aggregation/_moco.py:130]: The size of matrix/y are the critical part. This is slightly costly because we have another time the full Jacobian in memory. I would not use `y` and use `self._y` instead to garbage collect the previous value and save one full Jacobian. Actualy the most memory efficient implementation would be:
```suggestion
            self._y = (1-\beta_step) * self._y
            self._y += beta_step * matrix.detach()
```
With this you have exactly two Jacobians stored.

File: src/torchjd/aggregation/_moco.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+            yy_t = y @ y.T
+            if self.rho != 0.0:
+                eye = torch.eye(matrix.shape[0], device=matrix.device, dtype=matrix.dtype)
+                yy_t = yy_t + self.rho * eye
+            lambd = torch.softmax(lambd - gamma_step * (yy_t @ lambd), dim=-1)
+
+        self._y = y
+        self._lambd = lambd
+
+        return lambd
```
Comment (PierreQuinton) [inline src/torchjd/aggregation/_moco.py:140]: I think this is incorrect. In equation 10 of the paper, they use `lambda @ y` as an update, not `lambda @ matrix`. I think this would make moco not weighted (the row spans of `matrix` and `y` can be very different in general).

File: src/torchjd/aggregation/_moco.py [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+
+    @property
+    def y(self) -> Tensor:
+        if self._y is None:
+            raise RuntimeError("The moving gradient estimate is not initialized yet.")
+        return self._y
+
+    def _ensure_state(self, matrix: Matrix) -> None:
```
Comment (PierreQuinton) [inline src/torchjd/aggregation/_moco.py:148]: Is the role of this function to initialize the state if none? we call it differently if so, maybe something in the direction of `_conditionally_initialize_state` (but probably improvable).

File: src/torchjd/aggregation/_moco.py [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+            gamma=gamma,
+            gamma_sigma=gamma_sigma,
+            rho=rho,
+        )
+        self.register_full_backward_pre_hook(raise_non_differentiable_error)
+
+    def forward(self, matrix: Matrix, /) -> Tensor:
+        weights = self.weighting(matrix)
+        if matrix.shape[0] == 0:
+            return matrix.sum(dim=0)
```
Comment (PierreQuinton) [inline src/torchjd/aggregation/_moco.py:202]: can remove
```suggestion
```

File: src/torchjd/aggregation/_moco.py (hunk 5 above)
Comment (PierreQuinton) [reply src/torchjd/aggregation/_moco.py:140]: My bad, I just saw below that while this is a weighting, the moco aggregator is not weighted. I assumed wrongly that it was.

File: src/torchjd/aggregation/_moco.py [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+
+        vector = weights @ self.weighting.y
+        if matrix.requires_grad:
+            vector = vector + 0.0 * matrix.sum(dim=0)
```
Comment (PierreQuinton) [inline src/torchjd/aggregation/_moco.py:206]: I would not do that. No reason to create a wrong differentiation graph for vector. Maybe we want to make this aggregator/weighting pair non-differentiable (because we don't want to keep all the Jacobians for all times). This can be done after #677 is merged.

```suggestion
```

**Review by PierreQuinton — CHANGES_REQUESTED** [review]
Few things to change/clarify after review of this and the paper. Tests need a more thorough review.

File: src/torchjd/aggregation/_moco.py (hunk 4 above)
Comment (PierreQuinton) [reply src/torchjd/aggregation/_moco.py:130]: I think this will be complicated to adapt with the normalization wrapper as we normalize the `y` itself, not `matrix`. So I think the normalization step of equation 6 needs to be implemented here. Note that what opencode suggested above from LibMTL doesn't match what is said in the paper, so I'm not sure which one to use.

**Comment by PierreQuinton** [comment]
/opencode:Plan Can you review the tests? compare to the tests of the aggregation codebase and try to match the quality and reason about all concepts that needs to be tested.

**Comment by PierreQuinton** [comment]
From the opencode review please do:
- 1
- 3
- 4:  It is not stochastic, but still it would be good to have reproducibility, so if we run it twice with the same sequence of matrices, and a reset in between, it should return the same results.

For 5, no need, the properties are not satisfied (permutation invariance could be changed to permuting the sequence of all inputs to the aggregator with the same permutation, in which case it would become permutation invariant @ValerianRey )

File: src/torchjd/aggregation/_moco.py (hunk 2 above)
Comment (ValerianRey) [reply src/torchjd/aggregation/_moco.py:129]: It's now merged FYI.

**Comment by PierreQuinton** [comment]
After discussion with @ValerianRey we believe that the Weighting should be private. The reason is that it is not intended to be used alone, and also it is not usable with autogram. @rkhosrowshahi can you remove it from docs, __init__.py and mark it as protected?

**Comment by PierreQuinton** [comment]
Actually now that I think of it, I think we can improve on the structure of this. If you think about the mapping from `y` to `lambda`, then this is a Weighting of PSD matrices (it takes `yy_t` and outputs `lambd` based on it. So I think it makes sense implement it as such. Note that it would also be stateful, but keeps in memory just the lambdas. from there the wrapping class saves the `y`, updates it with `y - beta_step * (y - J)` and gives it to the wrapped Gramian weighting. I think it separates responsibilities a bit better (even though I still don't understand what they do nor why they do it). What do you think @rkhosrowshahi @ValerianRey ?

**Comment by ValerianRey** [comment]
> Actually now that I think of it, I think we can improve on the structure of this. If you think about the mapping from `y` to `lambda`, then this is a Weighting of PSD matrices (it takes `yy_t` and outputs `lambd` based on it. So I think it makes sense implement it as such. Note that it would also be stateful, but keeps in memory just the lambdas. from there the wrapping class saves the `y`, updates it with `y - beta_step * (y - J)` and gives it to the wrapped Gramian weighting. I think it separates responsibilities a bit better (even though I still don't understand what they do nor why they do it). What do you think @rkhosrowshahi @ValerianRey ?

I don't like so much have extra classes if those classes are gonna be private and not reusable anywhere else (it leads to more code, duplicated documentation of params, etc).

I even think we could merge the current two classes into just an aggregator, because the weighting won't be usable by itself anyway.

What do you think?

File: src/torchjd/aggregation/_moco.py (hunk 3 above)
Comment (PierreQuinton) [reply src/torchjd/aggregation/_moco.py:135]: @rkhosrowshahi @KhusPatel4450 This is common code to MOCO and MODO. We should extract it (note that we also need to resolve this issue of projection versus softmax) as a stateful gramian weighting that stores `lambd` and is parametrized by `gamma_step`. I'm not sure how to call this weighting.
